docs/pallas/tpu/matmul.md

```diff
@@ -0,0 +1,643 @@
+
+```{code-cell} ipython3
+:id: OpU7I7BNXQYg
+:outputId: 28c2c3cf-759e-465c-f969-0e2c9607b8a5
+
+print("================ XLA matmul ===================")
+mm = jnp.matmul
+analyze_matmul(1024, 1024, 1024, jnp.bfloat16, mm)
+analyze_matmul(4096, 4096, 4096, jnp.bfloat16, mm)
+analyze_matmul(8192, 8192, 8192, jnp.bfloat16, mm)
+```
+
++++ {"id": "L-KUG3lha-jm"}
+
+Pallas, with some very basic tuning, gets pretty close to XLA's performance numbers! By trying out more block sizes, we should expect to close the gap entirely.
+
++++ {"id": "nbdHMJRObnZa"}
+
+## Templating the matrix multiplication
+
++++ {"id": "qSfcMwtDg7Vn"}
+
+Now that we have a basic matrix multiplication kernel, we can now try fusing operations into it.
+
+### Fused right-hand-side transpose
+
+A common first thing to do is to fuse a transpose (e.g. instead of doing `x @ y` we do `x @ y.T`). On TPUs, the MXU supports matmul routines with the right hand side transposed natively, so there is no additional cost to fusing in a transpose. We can use this routine with `jax.lax.dot_general`.
+
+```{code-cell} ipython3
+:id: 1_6S_QnMbHAQ
+
+def matmul_kernel(x_ref, y_ref, z_ref, acc_ref, *, nsteps, transpose_rhs):
+  @pl.when(pl.program_id(2) == 0)
+  def _():
+    acc_ref[...] = jnp.zeros_like(acc_ref)
+
+  if transpose_rhs:
+    dims = ((1,), (1,)), ((), ())
+  else:
+    dims = ((1,), (0,)), ((), ())
+
+  acc_ref[...] += jax.lax.dot_general(
+      x_ref[...], y_ref[...], dims, preferred_element_type=jnp.float32,
+  )
+
+  @pl.when(pl.program_id(2) == nsteps - 1)
+  def _():
+    z_ref[...] = acc_ref[...].astype(z_ref.dtype)
+
+
+@functools.partial(jax.jit, static_argnames=['bm', 'bk', 'bn', 'transpose_rhs'])
+def matmul(
+    x: jax.Array,
+    y: jax.Array,
+    *,
+    bm: int = 128,
+    bk: int = 128,
+    bn: int = 128,
+    transpose_rhs: bool = False,
+):
+  if transpose_rhs:
+    y = y.swapaxes(0, 1)
+    y_block_spec = pl.BlockSpec(lambda i, j, k: (j, k), (bn, bk))
+  else:
+    y_block_spec = pl.BlockSpec(lambda i, j, k: (k, j), (bk, bn))
+  m, k = x.shape
+  _, n = y.shape
+  return pl.pallas_call(
+      functools.partial(matmul_kernel, nsteps=k // bk, transpose_rhs=transpose_rhs),
+      grid_spec=pltpu.PrefetchScalarGridSpec(
+        num_scalar_prefetch=0,
+        in_specs=[
+            pl.BlockSpec(lambda i, j, k: (i, k), (bm, bk)),
+            y_block_spec,
+        ],
+        out_specs=pl.BlockSpec(lambda i, j, k: (i, j), (bm, bn)),
+        scratch_shapes=[pltpu.VMEM((bm, bn), jnp.float32)],
+        grid=(m // bm, n // bn, k // bk),
+      ),
+      out_shape=jax.ShapeDtypeStruct((m, n), x.dtype),
+  )(x, y)
+```
+
++++ {"id": "eSmPJHSchuGX"}
+
+Note that we do a transpose inside of the `matmul` function (`y = y.swapaxes(0, 1)`). This is because transposes in XLA are *logical*, not physical. It informs XLA that the custom call would like `y` to be laid out differently before the matmul.
```

Comment:
I think this could use a clarification. For XLA, dimension ordering is purely logical and it can rearrange them. But, for Pallas kernels the data layout will always be enforced to be row-major. So by doing the transpose outside of the kernel we change the data layout enforced by the Pallas kernel.

It would also be worth saying that XLA can often elide the transpose by ensuring that the producer of the value already writes it out in the way the Pallas kernel wants to consume it.